### PR TITLE
More rp2040 BufferedUart fixes

### PR DESCRIPTION
**Pull request by timokroeger**
* Refactor init code
* Make it possible to drop RX without breaking TX (or vice versa)
* Correctly handle RX buffer full scenario

### CODE REVIEW

File: embassy-rp/src/uart/buffered.rs [hunk 1]
```diff
@@ -367,40 +305,31 @@ pub(crate) unsafe fn on_interrupt<T: Instance>(_: *mut ()) {
...
-            });
         }
         if ris.oeris() {
             warn!("Overrun error");
-            r.uarticr().write(|w| {
-                w.set_oeic(true);
-            });
         }
+        // Clear any error flags
+        r.uarticr().write(|w| {
```
Comment (Dirbaio) [inline embassy-rp/src/uart/buffered.rs:325]: if an error happens between reading `uartimsc` and this write, it'll get lost.

File: embassy-rp/src/uart/buffered.rs [hunk 2]
```diff
@@ -72,66 +100,25 @@ impl<'d, T: Instance> BufferedUart<'d, T> {
...
-        let state = T::state();
-        let regs = T::regs();
-
-        let len = tx_buffer.len();
-        unsafe { state.tx_buf.init(tx_buffer.as_mut_ptr(), len) };
-        let len = rx_buffer.len();
-        unsafe { state.rx_buf.init(rx_buffer.as_mut_ptr(), len) };
-
-        unsafe {
-            regs.uartimsc().modify(|w| {
```
Comment (Dirbaio) [inline embassy-rp/src/uart/buffered.rs:116]: this `uartimsc` write is gone in the new code. I think it's still necessary.

File: embassy-rp/src/uart/buffered.rs [hunk 3]
```diff
@@ -155,48 +143,24 @@ impl<'d, T: Instance> BufferedUartRx<'d, T> {
...
-        irq.set_handler(on_interrupt::<T>);
-        irq.unpend();
-        irq.enable();
-
         Self { phantom: PhantomData }
     }
 
     fn read<'a>(buf: &'a mut [u8]) -> impl Future<Output = Result<usize, Error>> + 'a {
         poll_fn(move |cx| {
+            unsafe { T::Interrupt::steal() }.pend();
```
Comment (Dirbaio) [inline embassy-rp/src/uart/buffered.rs:162]: We shouldn't always pend the irq, it wastes cpu time. It should be pended only if we we actually popped bytes off the buffer. Or if we want to be more precise, only when we popped bytes which caused the buffer to go from full to not-full.

Or alternatively, on pop enable the rx irq, and let the hardware trigger the irq when data actually arrives.

Same for `fill_buf` and `write`.

File: embassy-rp/src/uart/buffered.rs (hunk 1 above)
Comment (timokroeger) [reply embassy-rp/src/uart/buffered.rs]: fixed by only clearing the pending flags

**Review by Dirbaio — APPROVED** [review]
Thanks!

bors r+

**Pull request merged**